File: tests/fixtures/daemon_owner.py
#!/usr/bin/env python3
import os
import subprocess
import sys
import tempfile
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))
from tests.test_helper import process_cleanup

home = Path(tempfile.mkdtemp(prefix="interrupt-python-")) / "home"
mode, output = sys.argv[1], Path(sys.argv[2])
if mode == "direct":
    process = subprocess.Popen([sys.executable, str(ROOT / "bin/rozorod.py"), "--home", str(home)])
    process_cleanup.register(process, home)
else:
    home.mkdir(mode=0o700)
    subprocess.run([sys.executable, str(ROOT / "bin/rzr-monitor.py"), "start"],
                   env={**os.environ, "ROZORO_HOME": str(home)}, check=True)
    process_cleanup.register_lock(home)
while not (home / "monitor.sock").exists(): time.sleep(.01)
output.write_text(str(home))
if mode == "assertion":
    raise AssertionError("intentional interrupted-path assertion")
while True: time.sleep(1)

File: tests/fixtures/daemon-owner.ts
import { mkdir, writeFile } from "node:fs/promises";
import { spawn } from "node:child_process";
import { join } from "node:path";

const [home, output] = process.argv.slice(2);
await mkdir(home, {recursive:true, mode:0o700});
const daemon = spawn("python3", [join(process.cwd(), "bin/rozorod.py"), "--home", home], {stdio:"ignore"});
let cleaning = false;
const cleanup = async (signal: "SIGINT"|"SIGTERM") => {
  if (cleaning) return; cleaning = true;
  if (daemon.exitCode === null) {
    daemon.kill("SIGTERM");
    await new Promise<void>((resolve) => daemon.once("exit", () => resolve()));
  }
  const { rm } = await import("node:fs/promises"); await rm(home, {recursive:true, force:true});
  process.kill(process.pid, signal);
};
process.once("SIGINT", () => void cleanup("SIGINT"));
process.once("SIGTERM", () => void cleanup("SIGTERM"));
while (true) {
  try { const { lstat } = await import("node:fs/promises"); if ((await lstat(join(home,"monitor.sock"))).isSocket()) break; } catch {}
  await new Promise((resolve) => setTimeout(resolve, 10));
}
await writeFile(output, home);
await new Promise(() => {});
